### PR TITLE
Add new texture filtering mode "Auto Max Quality"

### LINKED ISSUE #13888: Add a setting for mipmapping
please add mipmaping in game setting like ppsspp v 1.4.1

### PR DESCRIPTION
**Pull request by hrydgard**
When enabled, tweaks texture filtering for best quality (in a huge majority of games, at least).

It does this by enforcing mipmapping and minification filtering, enforces anisotropic filtering for all mip selection modes (if that's separately enabled) and (if the backend supports it, currently Vulkan) autogenerates mipmaps.

This looks nice and flicker free in most games without any additional tweaking, including GTA and Burnout which have long been painfully flickery in the distance due to undersampling.

Needs a bit more testing before merge, and will add mip gen to more APIs.

Also, maybe this should have a different name, though "Auto Max Quality" is at least pretty descriptive.

Fixes #13888

### CODE REVIEW

File: GPU/Vulkan/TextureCacheVulkan.cpp [hunk 1]
```diff
@@ -751,6 +751,15 @@ void TextureCacheVulkan::BuildTexture(TexCacheEntry *const entry) {
 	// such as when using replacement textures - but let's keep the same amount of levels.
 	int maxLevelToGenerate = maxLevel;
 
+	if (g_Config.iTexFiltering == TEX_FILTER_AUTO_MAX_QUALITY) {
+		// Boost the number of mipmaps.
+		int maxPossibleMipmaps = log2i(std::min(gstate.getTextureWidth(0), gstate.getTextureHeight(0)));
+		if (maxPossibleMipmaps != maxLevelToGenerate) {
+			maxLevelToGenerate = maxPossibleMipmaps;
+		}
```
Comment (iota97) [inline GPU/Vulkan/TextureCacheVulkan.cpp:759]: nit: the check is not needed :)

**Comment by XiaomiRedmi9** [comment]
Texture MAX Size is also a good name?

**Comment by Narugakuruga** [comment]
Test result are very impressive.
some UI stuff are unimplemented. If change the texture filter in Windows menu, it give an error.

File: GPU/Vulkan/TextureCacheVulkan.cpp (hunk 1 above)
Comment (hrydgard) [reply GPU/Vulkan/TextureCacheVulkan.cpp:759]: Oh yeah, I only had it so I could set a breakpoint there, while debugging it.

File: Windows/MainWindowMenu.cpp [hunk 2]
```diff
@@ -1225,11 +1225,12 @@ namespace MainWindow {
 			ID_OPTIONS_TEXTUREFILTERING_AUTO,
 			ID_OPTIONS_NEARESTFILTERING,
 			ID_OPTIONS_LINEARFILTERING,
+			ID_OPTIONS_AUTOMAXQUALITYFILTERING,
```
Comment (unknownbrackets) [inline Windows/MainWindowMenu.cpp:1228]: Needs a `TranslateMenuItem`, and also a case.

-[Unknown]

**Review by unknownbrackets — APPROVED** [review]
Makes sense.  I think the only case where we wouldn't want to do this is when the mip levels are the same size, which is a different problem anyway.

Should we change the default?  A lot of games don't use mipmaps, but likely only for space reasons.  I haven't tried this yet, but I'd expect it to be a universal improvement.

-[Unknown]

**Comment by LunaMoo** [comment]
Didn't had time to try it yet, but how special cases like #6357 handled?

**Comment by unknownbrackets** [comment]
No worse than now.  Ultimately, because those cases (Tactics Ogre was another) don't have mip sizes of decreasing size, we detect that they're invalid for use as mip levels.  In those cases, I think we'll force mipmaps off.  So it won't be changed by this pull.

-[Unknown]

**Comment by hrydgard** [comment]
It's got positive effects even without the extra mip generation on the non-vulkan backends, so I'm gonna merge as is and add the mip generation later.

**Pull request merged**